bitvm/src/signatures/public.rs

```diff
@@ -0,0 +1,838 @@
+        let digit_signatures: Vec<[u8; 20]> = signature
+            .as_ref()
+            .iter()
+            .map(|digit_sig| std::array::from_fn(|i| digit_sig[i]))
+            .collect();
+        Self::CompactSignature::try_from(digit_signatures).unwrap()
+    }
+
+    /// Returns a Bitcoin script that verifies a Winternitz signature for the given `public_key`.
+    ///
+    /// ## Precondition
+    ///
+    /// Signature is at the stack top.
+    ///
+    /// ## Postcondition
+    ///
+    /// The converted message is on the stack top.
+    /// The checksum is consumed.
+    fn compact_checksig_verify(public_key: &Self::PublicKey) -> Script {
+        Self::COMPACT_ALGORITHM.checksig_verify(&Self::PARAMETERS, &public_key.as_ref().to_vec())
+    }
+
+    /// Returns a Bitcoin script that verifies a Winternitz signature for the given `public_key`.
+    ///
+    /// ## Precondition
+    ///
+    /// Signature (in the verifier's format) is at the stack top.
+    ///
+    /// ## Postcondition
+    ///
+    /// The message and checksum are consumed.
+    fn compact_checksig_verify_and_clear_stack(public_key: &Self::PublicKey) -> Script {
+        Self::COMPACT_ALGORITHM
+            .checksig_verify_and_clear_stack(&Self::PARAMETERS, &public_key.as_ref().to_vec())
+    }
+}
+
+/// Winternitz signatures for 4-byte messages.
+pub struct Wots4;
+/// Winternitz signatures for 16-byte messages.
+pub struct Wots16;
+/// Winternitz signatures for 32-byte messages.
+pub struct Wots32;
+/// Winternitz signatures for 64-byte messages.
+pub struct Wots64;
+/// Winternitz signatures for 72-byte messages.
+pub struct Wots72;
+/// Winternitz signatures for 80-byte messages.
+pub struct Wots80;
+
+macro_rules! impl_wots {
+    ($name:ident, $msg_byte_len:expr, $converter:ty) => {
+        impl Wots for $name {
+            type Converter = $converter;
+            /// The public key type for this Winternitz signing algorithm.
+            type PublicKey = [[u8; 20]; Self::TOTAL_DIGIT_LEN as usize];
+            /// The message type for this Winternitz signing algorithm.
+            ///
+            /// All messages have the same fixed length.
+            type Message = [u8; Self::MSG_BYTE_LEN as usize];
+            /// The signature type of this Winternitz signing algorithm.
+            type Signature = [[u8; 21]; Self::TOTAL_DIGIT_LEN as usize];
+
+            /// The number of bytes in a message.
+            const MSG_BYTE_LEN: u32 = $msg_byte_len;
+        }
+
+        impl CompactWots for $name {
+            /// The compact signature type of this Winternitz signing algorithm.
+            type CompactSignature = [[u8; 20]; Self::TOTAL_DIGIT_LEN as usize];
+        }
+    };
+}
+
+impl_wots!(Wots4, 4, VoidConverter);
+impl_wots!(Wots16, 16, VoidConverter);
+impl_wots!(Wots32, 32, VoidConverter);
+impl_wots!(Wots64, 64, VoidConverter);
+impl_wots!(Wots72, 72, ToBytesConverter);
+impl_wots!(Wots80, 80, VoidConverter);
```

Comment:
Is it worthwhile to include an explanatory comment to indicate the expected/existing use cases for these types?

Reply:
I will make the `impl_wots!` macro public, so you can generate the WOTS implementations that you need downstream.